1-Draft/RFCNNNN-Notification-On-Version-Update.md

```diff
@@ -0,0 +1,244 @@
+                todayFilePath,
+                FileMode.CreateNew,
+                FileAccess.Write,
+                FileShare.None,
+                bufferSize: 1024,
+                FileOptions.DeleteOnClose))
+        {
+            if (File.Exists(todayDoneFilePath))
+            {
+                // After grab the file lock, it turns out a successful check has finished.
+                // Then let's bail out early.
+                return;
+            }
+
+            // Now it's guaranteed that I'm the only process that reaches here.
+            foreach (string oldFile in Directory.EnumerateFiles(TestDir, "sentinel-*-*-*.done"))
+            {
+                // Clean up the old '.done' file, there should be only one.
+                File.Delete(oldFile);
+            }
+
+            // Do the real update check
+            //  - Send HTTP request to query for the new release/pre-release;
+            //  - If there is a valid new release that should be reported to the user, create the file `_update_<new-version>`,
+            //    or rename the existing `_update_<old-version>` to `_update_<new-version>`.
+            // ... more ...
+
+            // Finally, create the `todayDoneFilePath` file as an indicator that a successful update check has finished today.
+            new FileStream(
+                todayDoneFilePath,
+                FileMode.CreateNew,
+                FileAccess.Write,
+                FileShare.None,
+                bufferSize: 1024,
+                FileOptions.None).Close();
+        }
+    }
+    catch (Exception e)
+    {
+        // An update check is in progress from another `pwsh` process. So it's OK to just return.
+    }
+}
+```
+
+With the file lock, only one process can get in the guarded `using` block at a given time.
+So only one process will be creating the file `_update_<version>`, or renaming an old such file to reflect the new version.
+Yes, other processes could be looking at the old file name (when a `pwsh` session tries to print a notification),
+or working with an outdated `FileInfo` object (another update check tries to do a pre-validation).
+But it's fine for that to happen:
+
+- In the former case, that particular `pwsh` session will show a notification about an outdated version,
+  but the next `pwsh` session will show the right notification.
+- In the latter case, the other update check will continue and find the `.done` file already exists for today.
+
+It's possible that a `pwsh` session terminates while the update check task is still running,
+in the middle of the `using` block for example.
+Creating the `.done` file is the very last step in the `using` block.
+So if the session ends before the `.done` file is created,
+another update check will happen when the next `pwsh` session starts and finish the work.
+
+#### How to do the update check
+
+This is the easy part.
+
+- Determine if we need to check pre-releases.
+- Send HTTP query request.
```

Comment:
Do we store the version in a well-known Azure blob?

Reply:
That would be convenient for querying.
It's OK that we just depend on the GitHub rest APIs. I thought about the throttling limit of GitHub APIs, but it should be fine because for a particular version of `pwsh`, only a few requests will be send per day (ideally, only one HTTP request is sent, but it's possible that the process is terminated before it writes the `.done` file, so another process will pick up the work and query again).

Reply:
I was thinking it would be a public well-known URL so anyone can query it for their own purposes

Reply:
Sure, we can do that as part of our release

Reply:
Updated the RFC to mention that a well-known URL for latest release/pre-release information probably is needed.